lib/dry/validation/schema.rb

```diff
@@ -92,6 +94,14 @@ def [](name)
       def predicates
         self.class.predicates
       end
+
+      private
+
+      def validate_input(input)
+        unless input.is_a?(Hash)
+          raise ArgumentError, "expected a Hash but received +#{input.inspect}+"
+        end
```

Comment:
@solnic, similar errors are received when you pass in anything that doesn't respond to `[]` from `Validation.Result(input[name])` in `Dry::Validation::Rule::Key`. 

I saw that this would affect the other open pull request, so I thought we could use duck typing to verify the input. However, since that hasn't been pulled in yet, I figured it would be best to discuss the proper validation for future implementations.